extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js

```diff
@@ -246,26 +261,152 @@ export class AmpAdNetworkAdsenseImpl extends AmpA4A {
-          this.autoFormat_,
-          viewportSize,
-          this.element
-        ),
-        viewportSize.width
-      ).catch(() => {});
+    let upgradePromise = Promise.resolve();
+    // Note: this must happen before this.autoFormat_ is populated
+    // since it may change the outcome.
+    if (this.isInAdSizeOptimizationExperimentBranch()) {
+      upgradePromise = this.maybeUpgradeToResponsive();
     }
-    // This should happen last, as some diversion criteria rely on some of the
-    // preceding logic (specifically responsive logic).
-    this.divertExperiments();
-    this.maybeAddSinglePassExperiment();
+
+    return upgradePromise.then(() => {
+      this.autoFormat_ = this.element.getAttribute('data-auto-format') || '';
+
+      if (this.isResponsive_()) {
+        // Attempt to resize to the correct height. The width should already be
+        // 100vw, but is fixed here so that future resizes of the viewport don't
+        // affect it.
+        const viewportSize = this.getViewport().getSize();
+        return this.attemptChangeSize(
+          AmpAdNetworkAdsenseImpl.getResponsiveHeightForContext_(
+            this.autoFormat_,
+            viewportSize,
+            this.element
+          ),
+          viewportSize.width
+        ).catch(() => {});
+      }
+
+      // This should happen last, as some diversion criteria rely on some of the
+      // preceding logic (specifically responsive logic).
+      this.divertExperiments();
+      this.maybeAddSinglePassExperiment();
+    });
+  }
+
+  /**
+   * Upgrades the ad unit to responsive if there is an opt-in setting in localstorage.
+   * See https://github.com/ampproject/amphtml/issues/23568 for design.
+   * @return {Promise} a promise that resolves when any upgrade is complete.
+   */
+  maybeUpgradeToResponsive() {
+    // If the ad unit is already responsive, there's nothing to do.
+    // Note that we can't use this.autoFormat_ here because it's not
+    // populated yet.
+    if (this.element.hasAttribute('data-auto-format')) {
+      return Promise.resolve();
+    }
+    return (
+      Services.storageForDoc(this.element)
+        .then(storage => storage.get(this.getAdSizeOptimizationStorageKey()))
+        .then(isAdSizeOptimizationEnabled => {
+          if (isAdSizeOptimizationEnabled) {
+            this.upgradeToResponsive();
+          }
+        })
+        // Do nothing if we fail to read localstorage.
+        .catch(() => {
+          dev().warn(TAG, 'Failed to look up publisher ad size settings.');
+        })
+    );
+  }
+
+  /**
+   * Actually upgrades the ad unit to responsive.
+   */
+  upgradeToResponsive() {
+    this.element.setAttribute('height', ADSENSE_RSPV_WHITELISTED_HEIGHT);
+    this.element.setAttribute('width', '100vw');
+    this.element.setAttribute('data-full-width', '');
+    this.element.setAttribute('data-auto-format', 'rspv');
+
+    devAssert(this.isValidElement(), 'Upgraded element is not valid.');
```

Comment:
Why would this assert fail?  What about the change could make it invalid?

Reply:
For example if we didn't set height/width to the exact values we do above.

Reply:
devAssert will be stripped out from production binaries. is that what you want?

Reply:
Yes. This is intended to prevent accidentally breaking the feature during future development.